peercrawler.py

```diff
@@ -56,21 +57,32 @@ def __init__(self, ctx,
             thread = threading.Thread(target=self.run_periodic_cleanup, args=(inactivity_threshold_seconds,), daemon=True)
             thread.start()
 
-    def add_peers(self, new_peers: Iterable[Peer]):
+    def add_peers(self, from_peer: Peer, new_peers: Iterable[Peer]):
         with self.mutex:
-            new_peers = filter(lambda p: not p.ip.ipv6.is_unspecified, new_peers)  # filter out unspecified addresses
-            self.peers.update(new_peers)
+            for new_peer in new_peers:
+                if new_peer.ip.ipv6.is_unspecified:
+                    continue
+
+                if new_peer not in self.__connections_graph:
+                    self.__connections_graph[new_peer] = peer_set()
+
+                self.__connections_graph[from_peer].add(new_peer)
 
     def run_periodic_cleanup(self, inactivity_threshold_seconds):
         while True:
             with self.mutex:
-                self.peers.cleanup_inactive(inactivity_threshold_seconds)
+                for peer_collection in self.__connections_graph.values():
+                    peer_collection.cleanup_inactive(inactivity_threshold_seconds)
 
             time.sleep(inactivity_threshold_seconds)
 
-    def get_peers_copy(self):
+    def get_peers_copy(self) -> Collection[Peer]:
```

Comment:
This function name isn't correct anymore.
It does not return a copy of the peers.
It returns a different view into the peers.
We could rename it to `get_peers_as_list()` or `get_peers_copy_as_list`.

Reply:
You're right about that. The reason I decided to leave it as it is was because other modules may be using this method already and I didn't want to break the interface. I'll see if PyCharm can find out all the usages.

Reply:
You just need to grep for get_peers_copy